services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java

```diff
@@ -99,160 +111,128 @@ private void uploadInParts(PutObjectRequest putObjectRequest, long contentLength
-                                                           partNumber.getAndIncrement(),
-                                                           uploadId);
-
-            Consumer<CompletedPart> completedPartConsumer = completedPart -> completedParts.set(completedPart.partNumber() - 1,
-                                                                                                completedPart);
-            multipartUploadHelper.sendIndividualUploadPartRequest(uploadId, completedPartConsumer, futures,
-                                                                  Pair.of(uploadRequest, asyncRequestBody))
-                                 .whenComplete((r, t) -> {
-                                     if (t != null) {
-                                         if (failureActionInitiated.compareAndSet(false, true)) {
-                                             multipartUploadHelper.failRequestsElegantly(futures, t, uploadId, returnFuture,
-                                                                                         putObjectRequest);
-                                         }
-                                     } else {
-                                         completeMultipartUploadIfFinish(asyncRequestBodyInFlight.decrementAndGet());
-                                     }
-                                 });
-            subscription.request(1);
-        }
+        private KnownContentLengthAsyncRequestBodySubscriber subscriber;
 
-        @Override
-        public void onError(Throwable t) {
-            log.debug(() -> "Received onError ", t);
-            if (failureActionInitiated.compareAndSet(false, true)) {
-                multipartUploadHelper.failRequestsElegantly(futures, t, uploadId, returnFuture, putObjectRequest);
-            }
+        private DefaultPausableUpload(KnownContentLengthAsyncRequestBodySubscriber subscriber) {
+            this.subscriber = subscriber;
         }
 
         @Override
-        public void onComplete() {
-            log.debug(() -> "Received onComplete()");
-            isDone = true;
-            completeMultipartUploadIfFinish(asyncRequestBodyInFlight.get());
-        }
-
-        private void completeMultipartUploadIfFinish(int requestsInFlight) {
-            if (isDone && requestsInFlight == 0 && completedMultipartInitiated.compareAndSet(false, true)) {
-                CompletedPart[] parts =
-                    IntStream.range(0, completedParts.length())
-                             .mapToObj(completedParts::get)
-                             .toArray(CompletedPart[]::new);
-                multipartUploadHelper.completeMultipartUpload(returnFuture, uploadId, parts, putObjectRequest);
-            }
+        public S3ResumeToken pause() {
+            return subscriber.pause();
         }
+    }
 
+    @SdkTestInternalApi
+    public static DefaultPausableUpload defaultPausableUpload(KnownContentLengthAsyncRequestBodySubscriber subscriber) {
```

Comment:
This is a bit code smell; we should try to find other non-intrusive way to mock dependencies such as dependency injection.

I see this method is used in `DefaultFileUploadTest`. Do we actually need to replace subscriber? Can we mock `PausableUpload` and verify the mocked pausable upload?